hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java

```diff
@@ -3471,6 +3500,8 @@ public Builder withIncrementalTableServiceEnabled(boolean incrementalTableServic
 
     protected void setDefaults() {
       writeConfig.setDefaultValue(MARKERS_TYPE, getDefaultMarkersType(engineType));
+      // set default for streaming writes to metadata table.
+      writeConfig.setDefaultValue(STREAMING_WRITES_TO_METADATA_TABLE, getDefaultForStreamingWritesToMetadataTable(engineType));
```

Comment:
let's move all the default values inference to the options's inference function, so that there is no need to maintain it in multiple places: engine type and table version.

Reply:
oh, I did not see any precedence to use engine type in infer function.
will give it a try.

Reply:
ok, this is bit tricky. due to the limitation w/ ConfigProperty infra, we can only define `noDefault()` only for `String` types. Here we are using boolean. if not, we have to parse the boolean everytime from the string value. 
Prefer to leave it as is.